showkase-component-impressions/src/main/java/com/vinaygaba/showkase_component_impressions/ComponentImpressions.kt

```diff
@@ -0,0 +1,145 @@
+ * that's visible on the screen and the bounds of the composable.
+ */
+@OptIn(FlowPreview::class)
+fun <T> Modifier.visibilityEvents(
+    key: T,
+    onVisibilityChanged: (ShowkaseVisibilityEvent<T>) -> Unit,
+) = composed {
+    val view = LocalView.current
+    val lifecycle = LocalLifecycleOwner.current.lifecycle
+    val scope = rememberCoroutineScope()
+    val visibilityEvent by rememberUpdatedState(newValue = onVisibilityChanged)
+    val impressionCollector = remember(key) { ImpressionCollector<T>(scope, visibilityEvent) }
+    var visibilityMetadata: VisibilityMetadata? by remember { mutableStateOf(null) }
+
+    registerLifecycleImpressionEvents(key, lifecycle, impressionCollector, visibilityMetadata)
+    registerDisposeImpressionEvents(key, impressionCollector, lifecycle)
+    collectImpressionEvents(key, impressionCollector, visibilityEvent)
+
+    onGloballyPositioned { layoutCoordinates ->
+        visibilityMetadata = layoutCoordinates.visibilityPercentage(view = view)
+        impressionCollector.onLayoutCoordinatesChanged(key, visibilityMetadata!!)
+    }
+}
+
+/**
+ * Used for handling the use case where a composable function is not in composition anymore i.e is
+ * invisible.
+ */
+@SuppressLint("ComposableNaming")
+@Composable
+private fun <T> registerDisposeImpressionEvents(
+    key: T,
+    impressionCollector: ImpressionCollector<T>,
+    lifecycle: Lifecycle
+) {
+    DisposableEffect(key, lifecycle) {
+        onDispose {
+            impressionCollector.publishDisposeEvent(
+                key,
+                defaultVisibilityMetadata,
+            )
+        }
+    }
+}
+
+/**
+ * Collects the visibility change events from the internal coroutine flow and invokes the
+ * onVisibilityEvent that the user passed.
+ */
+@SuppressLint("ComposableNaming")
+@FlowPreview
+@Composable
+private fun <T> collectImpressionEvents(
+    key: T,
+    impressionCollector: ImpressionCollector<T>,
+    onVisibilityEvent: (ShowkaseVisibilityEvent<T>) -> Unit
+) {
+    LaunchedEffect(key) {
+        impressionCollector.impressionEvents
+            .collect { impressionData ->
+                val impression = impressionData as ShowkaseVisibilityEvent<*>
+                onVisibilityEvent(
+                    ShowkaseVisibilityEvent(
+                        impression.key as T,
+                        impression.visibilityPercentage,
+                        impression.boundsInWindow
+                    )
+                )
+            }
+    }
+}
+
+/**
+ * Used for handling the use case where the activity/app is backgrounded or foregrounded.
+ */
+@SuppressLint("RememberReturnType", "ComposableNaming")
+@Composable
+private fun <T> registerLifecycleImpressionEvents(
+    key: T,
+    lifecycle: Lifecycle,
+    impressionCollector: ImpressionCollector<T>,
+    visibility: VisibilityMetadata?,
+) {
+    remember(key1 = key, key2 = visibility) {
+        lifecycle.addObserver(
```

Comment:
where is this observer removed? don't we need to remove it as a dispose effect? seems like this could be in the same `DisposableEffect(key, lifecycle)` block used above